File: backend/src/repositories/professionalRepository.js
const Professional = require('../models/Professional');

class ProfessionalRepository {
  async create(professionalData) {
    try {
      const professional = new Professional(professionalData);
      return await professional.save();
    } catch (error) {
      throw error;
    }
  }

  async findByUserId(userId) {
    try {
      return await Professional.findOne({ userId });
    } catch (error) {
      throw error;
    }
  }

  async findById(id) {
    try {
      return await Professional.findById(id).populate('userId');
    } catch (error) {
      throw error;
    }
  }

  async update(id, updates) {
    try {
      return await Professional.findByIdAndUpdate(
        id,
        { $set: updates },
        { new: true, runValidators: true }
      );
    } catch (error) {
      throw error;
    }
  }

  async deleteByUserId(userId) {
    try {
      return await Professional.deleteOne({ userId });
    } catch (error) {
      throw error;
    }
  }

  async getFeatured(limit = 10) {
    try {
      return await Professional.find()
        .sort({ rating: -1 })
        .limit(limit)
        .populate('userId');
    } catch (error) {
      throw error;
    }
  }

  async search({ query, category, minRating, maxPrice, location, limit = 20, skip = 0 }) {
    try {
      const filter = {};

      if (category) {
        filter.categories = category;
      }

      if (minRating !== undefined) {
        filter.rating = { $gte: minRating };
      }

      if (location) {
        filter.coverageArea = { $regex: location, $options: 'i' };
      }

      let professionals = await Professional.find(filter)
        .limit(limit)
        .skip(skip)
        .populate('userId');

      if (query) {
        const queryLower = query.toLowerCase();
        professionals = professionals.filter(p => {
          const user = p.userId;
          if (!user) return false;
          
          const nameMatch = user.name?.toLowerCase().includes(queryLower);
          const bioMatch = p.bio?.toLowerCase().includes(queryLower);
          const categoryMatch = p.categories?.some(c => 
            c.toLowerCase().includes(queryLower)
          );
          
          return nameMatch || bioMatch || categoryMatch;
        });
      }

      if (maxPrice !== undefined) {
        professionals = professionals.filter(p => 
          !p.hourlyRate || p.hourlyRate <= maxPrice
        );
      }

      return professionals;
    } catch (error) {
      throw error;
    }
  }

  async count(filters = {}) {
    try {
      const filter = {};

      if (filters.category) {
        filter.categories = filters.category;
      }

      if (filters.minRating !== undefined) {
        filter.rating = { $gte: filters.minRating };
      }

      if (filters.location) {
        filter.coverageArea = { $regex: filters.location, $options: 'i' };
      }

      // Para query e maxPrice, precisamos buscar e filtrar em memória
      if (filters.query || filters.maxPrice !== undefined) {
        let professionals = await Professional.find(filter).populate('userId');
        
        if (filters.query) {
          const queryLower = filters.query.toLowerCase();
          professionals = professionals.filter(p => {
            const user = p.userId;
            if (!user) return false;
            
            const nameMatch = user.name?.toLowerCase().includes(queryLower);
            const bioMatch = p.bio?.toLowerCase().includes(queryLower);
            const categoryMatch = p.categories?.some(c => 
              c.toLowerCase().includes(queryLower)
            );
            
            return nameMatch || bioMatch || categoryMatch;
          });
        }

        if (filters.maxPrice !== undefined) {
          professionals = professionals.filter(p => 
            !p.hourlyRate || p.hourlyRate <= filters.maxPrice
          );
        }

        return professionals.length;
      }

      return await Professional.countDocuments(filter);
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new ProfessionalRepository();

File: lib/services/repositories/professional_repository.dart
import '../../models/professional_model.dart';
import '../api/api_service.dart';

class ProfessionalRepository {
  static Future<String?> create(ProfessionalModel professional) async {
    final response = await ApiService.post('/professionals', professional.toJson());
    if (response['success'] == true && response['professional'] != null) {
      return response['professional']['id'] as String?;
    }
    return null;
  }

  static Future<ProfessionalModel?> findByUserId(String userId) async {
    final response = await ApiService.get('/professionals/user/$userId');
    if (response['success'] == true && response['professional'] != null) {
      return ProfessionalModel.fromJson(response['professional']);
    }
    return null;
  }

  static Future<Map<String, dynamic>?> findByIdWithUser(String id) async {
    final response = await ApiService.get('/professionals/$id');
    if (response['success'] == true) {
      return {
        'professional': response['professional'],
        'user': response['user'],
      };
    }
    return null;
  }

  static Future<ProfessionalModel?> findById(String id) async {
    final response = await ApiService.get('/professionals/$id');
    if (response['success'] == true && response['professional'] != null) {
      final professionalData = response['professional'] as Map<String, dynamic>;
      // Se houver userId no user, usar o ID do user
      if (response['user'] != null) {
        professionalData['userId'] = response['user']['id'];
      }
      return ProfessionalModel.fromJson(professionalData);
    }
    return null;
  }

  static Future<bool> update(String id, Map<String, dynamic> updates) async {
    final response = await ApiService.put('/professionals/$id', updates, requiresAuth: true);
    return response['success'] == true;
  }

  static Future<bool> deleteByUserId(String userId) async {
    final response = await ApiService.delete('/professionals/user/$userId', requiresAuth: true);
    return response['success'] == true;
  }

  static Future<List<ProfessionalModel>> getFeatured({int limit = 10}) async {
    final response = await ApiService.get('/professionals/featured?limit=$limit');
    if (response['success'] == true && response['results'] != null) {
      final results = response['results'] as List;
      return results.map((p) => ProfessionalModel.fromJson(p['professional'])).toList();
    }
    return [];
  }

  static Future<List<Map<String, dynamic>>> getFeaturedWithUserInfo({int limit = 10}) async {
    final response = await ApiService.get('/professionals/featured?limit=$limit');
    if (response['success'] == true && response['results'] != null) {
      return List<Map<String, dynamic>>.from(response['results']);
    }
    return [];
  }

  static Future<List<ProfessionalModel>> search({
    String? query,
    String? category,
    double? minRating,
    double? maxPrice,
    String? location,
    int limit = 20,
    int skip = 0,
  }) async {
    final queryParams = <String>[];
    if (query != null) queryParams.add('query=$query');
    if (category != null) queryParams.add('category=$category');
    if (minRating != null) queryParams.add('minRating=$minRating');
    if (maxPrice != null) queryParams.add('maxPrice=$maxPrice');
    if (location != null) queryParams.add('location=$location');
    queryParams.add('limit=$limit');
    queryParams.add('skip=$skip');

    final endpoint = '/professionals/search?${queryParams.join('&')}';
    final response = await ApiService.get(endpoint);
    
    if (response['success'] == true && response['results'] != null) {
      final results = response['results'] as List;
      return results.map((p) => ProfessionalModel.fromJson(p['professional'])).toList();
    }
    return [];
  }

  static Future<List<Map<String, dynamic>>> searchWithUserInfo({
    String? query,
    String? category,
    double? minRating,
    double? maxPrice,
    String? location,
    int limit = 20,
    int skip = 0,
  }) async {
    final queryParams = <String>[];
    if (query != null) queryParams.add('query=$query');
    if (category != null) queryParams.add('category=$category');
    if (minRating != null) queryParams.add('minRating=$minRating');
    if (maxPrice != null) queryParams.add('maxPrice=$maxPrice');
    if (location != null) queryParams.add('location=$location');
    queryParams.add('limit=$limit');
    queryParams.add('skip=$skip');

    final endpoint = '/professionals/search?${queryParams.join('&')}';
    final response = await ApiService.get(endpoint);
    
    if (response['success'] == true && response['results'] != null) {
      return List<Map<String, dynamic>>.from(response['results']);
    }
    return [];
  }

  static Future<int> count({
    String? query,
    String? category,
    double? minRating,
    double? maxPrice,
    String? location,
  }) async {
    final queryParams = <String>[];
    if (query != null) queryParams.add('query=$query');
    if (category != null) queryParams.add('category=$category');
    if (minRating != null) queryParams.add('minRating=$minRating');
    if (maxPrice != null) queryParams.add('maxPrice=$maxPrice');
    if (location != null) queryParams.add('location=$location');

    final endpoint = '/professionals/count?${queryParams.join('&')}';
    final response = await ApiService.get(endpoint);
    
    if (response['success'] == true && response['total'] != null) {
      return response['total'] as int;
    }
    return 0;
  }
}

